src/Symfony/Bridge/Doctrine/Tests/Security/User/EntityUserProviderTest.php

```diff
@@ -233,14 +233,44 @@ private function getManager($em, $name = null)
 
     private function getObjectManager($repository)
     {
-        $em = $this->getMockBuilder(ObjectManager::class)
-            ->onlyMethods(['getClassMetadata', 'getRepository'])
-            ->getMockForAbstractClass();
-        $em->expects($this->any())
-            ->method('getRepository')
-            ->willReturn($repository);
+        $objectManager = new class($repository) implements ObjectManager {
```

Comment:
How did you decide whether to use an anonymous classes, a mocks, or a stubs, for each file?

Reply:
I use anonymous classes when I need some logic from parent class to be executed. I personally think it's better to stay away from mocking when possible in order to stick to the "reality" as much as we can (and also avoid dependencies to mock frameworks).

Stubs can be used to mock things that are more complicated to implement but we don't need to do expectations on args that are passed, we just care about some return value. Like a "simpler mock".

Mock are used for the rest 🙂 

This is kinda my rule of thumb, even I can't tell if it is truly exact!